misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/BinaryExpressionEvaluator.java

```diff
@@ -387,6 +403,50 @@ private BExpressionValue conditionalReturn(BVariable lVar, BVariable rVar) {
         }
     }
 
+    /**
+     * Checks for deep value equality.
+     */
+    private BExpressionValue checkValueEquality(BVariable lVar, BVariable rVar, SyntaxKind operatorType)
+            throws EvaluationException {
+        List<Value> argList = new ArrayList<>();
+        argList.add(getValueAsObject(context, lVar));
+        argList.add(getValueAsObject(context, rVar));
+
+        List<String> argTypeNames = new ArrayList<>();
+        argTypeNames.add(JAVA_OBJECT_CLASS);
+        argTypeNames.add(JAVA_OBJECT_CLASS);
+        RuntimeStaticMethod runtimeMethod = getRuntimeMethod(context, B_TYPE_CHECKER_CLASS, VALUE_EQUAL_METHOD,
+                argTypeNames);
+        runtimeMethod.setArgValues(argList);
+        Value result = runtimeMethod.invoke();
+        BVariable variable = VariableFactory.getVariable(context, result);
+        boolean booleanValue = Boolean.parseBoolean(variable.getDapVariable().getValue());
+        booleanValue = operatorType == SyntaxKind.DOUBLE_EQUAL_TOKEN ? booleanValue : !booleanValue;
+        return EvaluationUtils.make(context, booleanValue);
+    }
+
+    /**
+     * Checks for reference equality.
+     */
+    private BExpressionValue checkReferenceEquality(BVariable lVar, BVariable rVar, SyntaxKind operatorType)
+            throws EvaluationException {
+        List<Value> argList = new ArrayList<>();
+        argList.add(getValueAsObject(context, lVar));
+        argList.add(getValueAsObject(context, rVar));
+
+        List<String> argTypeNames = new ArrayList<>();
+        argTypeNames.add(JAVA_OBJECT_CLASS);
+        argTypeNames.add(JAVA_OBJECT_CLASS);
```

Comment:
Same [here](https://github.com/ballerina-platform/ballerina-lang/pull/26322/files#r505441120) as well. Is there any specific reason for this?

Reply:
This is intentional. This relevant method accepts two java.lang.Object instances as its arguments.

Reply:
https://github.com/ballerina-platform/ballerina-lang/pull/26322#discussion_r505464641